nima/webserver/webserver/src/main/java/io/helidon/nima/webserver/ServerListener.java

```diff
@@ -361,16 +383,23 @@ private void listen() {
                     }
 
                     handler = new ConnectionHandler(this,
+                                                    connectionSemaphore,
+                                                    requestSemaphore,
                                                     connectionProviders,
+                                                    activeConnections,
                                                     helidonSocket,
                                                     router);
 
                     readerExecutor.execute(handler);
                 } catch (RejectedExecutionException e) {
                     LOGGER.log(ERROR, "Executor rejected handler for new connection");
+                    // we never started the handler, so we must release the semaphore here
+                    connectionSemaphore.release();
                 } catch (Exception e) {
                     // we may get an SSL handshake errors, which should only fail one socket, not the listener
                     LOGGER.log(TRACE, "Failed to handle accepted socket", e);
+                    // we never started the handler, so we must release the semaphore here
+                    connectionSemaphore.release();
```

Comment:
This covers the case when waiting acquire is interrupted by idle connection timer right?

Reply:
yes it should, as the exception would be caught here. I have tried to cover all "escape routes" from the code, and always release the semaphore.